sktime/utils/mlflow_sktime.py

```diff
@@ -0,0 +1,804 @@
+    internally to serialize a model.
+mlflow.pyfunc
+    Produced for use by generic pyfunc-based deployment tools and batch inference.
+
+    The `pyfunc` flavor of the model supports sktime predict methods `predict`,
+    `predict_interval`, `predict_proba`, `predict_quantiles`, `predict_var`.
+
+    The interface for utilizing a sktime model loaded as a `pyfunc` type for
+    generating forecasts requires passing an exogenous regressor as Pandas
+    DataFrame to the `pyfunc.predict()` method (an empty DataFrame must be
+    passed if no exogenous regressor is used). The configuration of predict
+    methods and parameter values passed to the predict methods is defined by
+    a dictionary to be saved as an attribute of the fitted sktime model
+    instance. If no prediction configuration is defined `pyfunc.predict()`
+    will return output from sktime `predict` method. Note that for `pyfunc`
+    flavor the forecasting horizon `fh` must be passed to the fit method.
+
+    Predict methods and parameter values for `pyfunc` flavor can be defined
+    in two ways: `Dict[str, dict]` if parameter values are passed to
+    `pyfunc.predict()`, for example
+    `{"predict_method": {"predict": {}, "predict_interval": {"coverage": [0.1, 0.9]}}`.
+    `Dict[str, list]`, with default parameters in predict method, for example
+    `{"predict_method": ["predict", "predict_interval"}` (Note: when including
+    `predict_proba` method the former appraoch must be followed as `quantiles`
+    parameter has to be provided by the user). If no prediction config is defined
+    `pyfunc.predict()` will return output from sktime `predict()` method.
+"""
+
+__author__ = ["benjaminbluhm"]
+__all__ = [
+    "get_default_pip_requirements",
+    "get_default_conda_env",
+    "save_model",
+    "log_model",
+    "load_model",
+]
+
+import logging
+import os
+import pickle
+
+import pandas as pd
+import yaml
+
+import sktime
+from sktime import utils
+from sktime.utils.multiindex import flatten_multiindex
+from sktime.utils.validation._dependencies import _check_soft_dependencies
+
+if _check_soft_dependencies("mlflow", severity="warning"):
+    from mlflow import pyfunc
+
+FLAVOR_NAME = "mlflow_sktime"
+
+PYFUNC_PREDICT_CONF = "pyfunc_predict_conf"
+PYFUNC_PREDICT_CONF_KEY = "predict_method"
+SKTIME_PREDICT = "predict"
+SKTIME_PREDICT_INTERVAL = "predict_interval"
+SKTIME_PREDICT_PROBA = "predict_proba"
+SKTIME_PREDICT_QUANTILES = "predict_quantiles"
+SKTIME_PREDICT_VAR = "predict_var"
+SUPPORTED_SKTIME_PREDICT_METHODS = [
+    SKTIME_PREDICT,
+    SKTIME_PREDICT_INTERVAL,
+    SKTIME_PREDICT_PROBA,
+    SKTIME_PREDICT_QUANTILES,
+    SKTIME_PREDICT_VAR,
+]
+
+SERIALIZATION_FORMAT_PICKLE = "pickle"
+SERIALIZATION_FORMAT_CLOUDPICKLE = "cloudpickle"
+SUPPORTED_SERIALIZATION_FORMATS = [
+    SERIALIZATION_FORMAT_PICKLE,
+    SERIALIZATION_FORMAT_CLOUDPICKLE,
+]
+
+_logger = logging.getLogger(__name__)
+
+
+def get_default_pip_requirements(include_cloudpickle=False):
```

Comment:
actually I am wondering if we can include the soft dependencies also here somehow? Is it possible to accept a `list` as input for `get_default_pip_requirements` and for `get_default_conda_env`

Reply:
not sure exactly what you mean, what would be the content of that `list`? and how does it relate to soft dependencies?

Reply:
I thought this 2 functions are used to restore an environment on which the forecaster was trained on? So if the forecaster used a soft-dependency e.g. `pmdarima` it could also be recreated with that version of `pmdarima`? So we could optionally handover a list of soft-dependencies to store them as well, not only the `cloudpickle`?

Reply:
Ah got you, I think for this purpose we could hand over a list of requirements to `pip_requirements` argument in `save_model` and `log_model` function and `_process_pip_requirements` would then combine these additional requirements with the default requirements `sktime` and `pickle` (or `cloudpickle`)

Reply:
yeah sth like that, but can also be done in a separate PR as it would be backwards compatible?